include/dali/core/boundary.h

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_BOUNDARY_H_
+#define DALI_CORE_BOUNDARY_H_
+
+#include "dali/core/host_dev.h"
+#include "dali/core/force_inline.h"
+#include "dali/core/geom/vec.h"
+
+namespace dali {
+
+/// Out-of-bounds index handling
+namespace boundary {
+
+/**
+ * @brief Reflects out-of-range indices until fits in range.
+ *
+ * @param idx The index to clamp
+ * @param lo Low (inclusive) bound
+ * @param hi High (exclusive) bound
+ *
+ * This reflect flavor does not repeat the first and last element:
+ * ```
+ * lo--v     v--hi
+ *     ABCDEF         is padded as
+ * EDCBABCDEFEDCBABCD
+ * ```
+ */
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE
+std::enable_if_t<std::is_integral<T>::value, T> idx_reflect_101(T idx, T lo, T hi) {
```

Comment:
Do we have use cases for this template with types allowed by is_integral other than int? Same goes for couple of other templates in this PR.